From 217207da194cdf9136fd4cee16d4ba6f57491f68 Mon Sep 17 00:00:00 2001
From: KokaKiwi <kokakiwi+git@kokakiwi.net>
Date: Sat, 11 Jun 2022 23:08:31 +0200
Subject: [PATCH 3/3] fix: Deduplicate resources directories

It seems to actually happens despite xdgpp already doing that...
---
 lib/libimhex/include/hex/helpers/utils.hpp | 11 +++++++++++
 lib/libimhex/source/helpers/fs.cpp         |  4 ++++
 2 files changed, 15 insertions(+)

diff --git a/lib/libimhex/include/hex/helpers/utils.hpp b/lib/libimhex/include/hex/helpers/utils.hpp
index 5f4e29af..26ff62df 100644
--- a/lib/libimhex/include/hex/helpers/utils.hpp
+++ b/lib/libimhex/include/hex/helpers/utils.hpp
@@ -234,6 +234,17 @@ namespace hex {
         return result;
     }
 
+    template<typename T>
+    void deduplicateVector(std::vector<T> &items) {
+        auto end = items.end();
+
+        for (auto it = items.begin(); it != end; ++it) {
+            end = std::remove(it + 1, end, *it);
+        }
+
+        items.erase(end, items.end());
+    }
+
     std::vector<std::string> splitString(const std::string &string, const std::string &delimiter);
     std::string combineStrings(const std::vector<std::string> &strings, const std::string &delimiter = "");
 
diff --git a/lib/libimhex/source/helpers/fs.cpp b/lib/libimhex/source/helpers/fs.cpp
index 392237af..53cce4c3 100644
--- a/lib/libimhex/source/helpers/fs.cpp
+++ b/lib/libimhex/source/helpers/fs.cpp
@@ -1,4 +1,5 @@
 #include <hex/helpers/fs.hpp>
+#include <hex/helpers/utils.hpp>
 
 #include <hex/api/content_registry.hpp>
 #include <hex/helpers/fs_macos.hpp>
@@ -256,6 +257,9 @@ namespace hex::fs {
         if (exePath && !exePath->empty())
             dataDirs.push_back(exePath->parent_path());
 
+        hex::deduplicateVector(configDirs);
+        hex::deduplicateVector(dataDirs);
+
         switch (path) {
             case ImHexPath::Patterns:
                 addUserDirs(dataDirs);
-- 
2.37.1

